src/main/java/bhs/devilbotz/commands/auto/DriveStraightToDock.java

```diff
@@ -4,95 +4,103 @@
-    drivetrain.arcadeDrive(speedSlewRateLimiter.calculate(output), -turnError);
+    double currentRoll = drivetrain.getRoll();
 
-    SmartDashboard.putNumber("Distance output", output);
-    SmartDashboard.putNumber("Position Tolerance", distancePid.getPositionTolerance());
-    SmartDashboard.putBoolean("at Setpoint", distancePid.atSetpoint());
-    SmartDashboard.putNumber("Position Error", distancePid.getPositionError());
-    SmartDashboard.putNumber("Average Distance", drivetrain.getAverageDistance());
-    SmartDashboard.putNumber("Left distance", drivetrain.getLeftDistance());
-    SmartDashboard.putNumber("Right Distance", drivetrain.getRightDistance());
+    switch (currentState) {
+      case ON_GROUND:
+        /* We start in the ON_GROUND state. We see if the currentRoll is within the expected ramp angle.
+         * Empirically, the roll is between 10 and 15 when on the ramp.
+         * When approaching the ramp, we cap the max speed to prevent crashing into the ramp
+         */
+        maxSpeed = 0.75;
+        if ((currentRoll > 10) && (currentRoll < 15)) {
+          onRampCount++;
+          System.out.println("#### Maybe On Ramp (onRampCount: " + onRampCount + ") ####");
+        } else {
+          onRampCount = 0;
+          System.out.println("#### Not on Ramp ####");
+        }
 
-    SmartDashboard.putNumber("Turn output", turnError);
-  }
+        /* If we've been on the ramp long enough (roll is within expected window), we assume we are on the ramp and transition states */
+        if (onRampCount > 10) {
+          System.out.println("#### On Ramp! ####");
+          currentState = DockState.ON_RAMP;
+        }
+        break;
 
-  // Called once the command ends or is interrupted.
-  @Override
-  public void end(boolean interrupted) {
-    drivetrain.arcadeDrive(0, 0);
-    System.out.println("DriveStraightToDock Finished");
+      case ON_RAMP:
+        /* When we think we are on the ramp, we reduce the maxSpeed so that we don't go so fast that we overshoot and cause the
+         * ramp to teeter quickly to the other side
+         * While the roll is decreasing and the currentRoll is less than the empirical ramp roll we assume we are leveling off.
+         */
+        maxSpeed = 0.50;
+        double deltaRoll = currentRoll - previousRoll;
+        if ((deltaRoll < 0) && (currentRoll < 10)) {
+          levelingRampCount++;
+          System.out.println(
+              "#### Maybe Leveling Off (levelingRampCount: " + levelingRampCount + ") ####");
+        } else {
+          levelingRampCount = 0;
+          System.out.println("#### Not Leveling Off ####");
+        }
+
+        /* If we've been leveling off long enough, we assume we are almost balanced */
+        if (levelingRampCount > 2) {
+          System.out.println("#### Leveling Off! ####");
+          currentState = DockState.LEVELING_OFF;
+        }
+        break;
+
+      case LEVELING_OFF:
+        maxSpeed = 0.01;
```

Comment:
What happens when setting maxSpeed to such a small value? Does the drivetrain actually move?

Reply:
This is somewhat of a hack.  When we detect that we are "LEVELING_OFF", the command actually exits.  See the isFinished() call. In the meantime we want to stop the robot from driving. In DriveStraightPID, maxSpeed = 0 means the value isn't set, so use the default robot speed (which is fast).  So, I'm just setting it to a very small value.  The command will exit immediately after the execute(), anyway.